x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/query/SparseTermsQuery.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    private static class SimpleLinearScorer extends Similarity.SimScorer {
+        private final float termValue;
+        private final float boost;
+
+        SimpleLinearScorer(float termValue, float boost) {
+            this.termValue = termValue;
+            this.boost = boost;
+        }
+
+        @Override
+        public float score(float freq, long norm) {
+            return decodeFeatureValue(freq) * termValue * boost;
+        }
+
+        @Override
+        public Explanation explain(Explanation freq, long norm) {
+            float score = score(freq.getValue().floatValue(), 1L);
+            return Explanation.match(
+                score,
+                "score(termValue=" + decodeFeatureValue(freq.getValue().floatValue()) + ", termQueryWeight=" + termValue + ")"
+            );
+        }
+    }
+
+    // Based on Lucene package private and final class TermWeight, but with specific scorer & optimizations
+    class SparseTermWeight extends Weight {
+        private final ScoreMode scoreMode;
+        private final Similarity.SimScorer simScorer;
+        private final TermStates termStates;
+
+        SparseTermWeight(ScoreMode scoreMode, float boost, TermStates termStates) {
+            super(SparseTermsQuery.this);
+            this.scoreMode = scoreMode;
+            this.simScorer = new SimpleLinearScorer(value, boost);
+            this.termStates = termStates;
+        }
+
+        @Override
+        public Matches matches(LeafReaderContext context, int doc) throws IOException {
+            TermsEnum te = getTermsEnum(context);
+            if (te == null) {
+                return null;
+            }
+            return MatchesUtils.forField(term.field(), () -> {
+                PostingsEnum pe = te.postings(null, PostingsEnum.OFFSETS);
+                if (pe.advance(doc) != doc) {
+                    return null;
+                }
+                return new TermMatchesIterator(getQuery(), pe);
+            });
+        }
+
+        @Override
+        public String toString() {
+            return "weight(" + SparseTermsQuery.this + ")";
+        }
+
+        @Override
+        public Scorer scorer(LeafReaderContext context) throws IOException {
+            final TermsEnum termsEnum = getTermsEnum(context);
+            if (termsEnum == null) {
+                return null;
+            }
+            LeafSimScorer scorer = new LeafSimScorer(simScorer, context.reader(), term.field(), scoreMode.needsScores());
+            if (scoreMode == ScoreMode.TOP_SCORES) {
+                return new TermScorer(this, termsEnum.impacts(PostingsEnum.FREQS), scorer);
+            } else {
+                return new TermScorer(
+                    this,
+                    termsEnum.postings(null, scoreMode.needsScores() ? PostingsEnum.FREQS : PostingsEnum.NONE),
+                    scorer
+                );
+            }
```

Comment:
FeatureField was written with the assumption that it would only make sense with scores, because it would be used to boost the hits of another query via a SHOULD clause, something like:

```
bool: [
  must: { "your main query goes here" },
  should: { "your rank_features query goes there" }
]
```

And if you'd run that query without scores, Lucene would ignore the `SHOULD` clause anyway.

Now that we're thinking of using this query in a more general way, IMO it makes sense to push your suggested changes to Lucene's FeatureQuery.